passes/opt/opt_expr.cc

```diff
@@ -1251,7 +1251,165 @@ void replace_const_cells(RTLIL::Design *design, RTLIL::Module *module, bool cons
 					}
 			}
 		}
-
+		
+		if (do_fine && (cell->type.in("$eq", "$ne")))
+		{
+			RTLIL::SigSpec a_sig, b_sig, y_sig;
+			
+			int a_width, b_width;
+			bool a_signed, b_signed;
+			
+			a_sig = assign_map(cell->getPort("\\A"));
+			b_sig = assign_map(cell->getPort("\\B"));
+			y_sig = assign_map(cell->getPort("\\Y"));
+			a_width = cell->parameters["\\A_WIDTH"].as_int();
+			b_width = cell->parameters["\\B_WIDTH"].as_int();
+			a_signed = cell->parameters["\\A_SIGNED"].as_bool();
+			b_signed = cell->parameters["\\B_SIGNED"].as_bool();
+			
+			dict<SigBit, RTLIL::SigSpec> compared_by_const;
+			pool<SigBit> signals_map;
+			SigBit polarity_bit = cell->type == "$eq" ? State::S0 : State::S1;
+			
+			bool do_gate = true;
+			
+			if (a_signed || b_signed)
+				do_gate = false;
+			
+			RTLIL::SigSpec all_input_bits;
+			all_input_bits.append(cell->getPort("\\A"));
+			all_input_bits.append(cell->getPort("\\B"));
+			
+			// skip cells that has x's or z's in inputs.
+			if (all_input_bits.to_sigbit_set().count(RTLIL::Sx) || all_input_bits.to_sigbit_set().count(RTLIL::Sz))
+				do_gate = false;
+			
+			if (do_gate)
+			{
+				if (GetSize(a_sig) < GetSize(b_sig))
+					a_sig.extend_u0(GetSize(b_sig), false);
+				if (GetSize(b_sig) < GetSize(a_sig))
+					b_sig.extend_u0(GetSize(a_sig), false);
+				
+				for (int i = 0; i < a_width; i++)
+				{
+					if ((a_sig[i] ==  RTLIL::State::S0) || (a_sig[i] ==  RTLIL::State::S1))
+					{
+						if (compared_by_const.count(b_sig[i]))
+						{
+							if (compared_by_const.at(b_sig[i]) != a_sig[i])
+							{
+								// case where comparing signal against S0 and S1
+								log("Replacing cell `%s' in module `%s' with zero-driver.\n",
+									cell->name.c_str(), module->name.c_str());
+								
+								module->connect(RTLIL::SigSig(y_sig, RTLIL::SigSpec(polarity_bit, y_sig.size())));
```

Comment:
When polarity_bit is "1" and y_sig.size() is >1 then this will set all bits, but only the LSB should be set.